<a href="https://colab.research.google.com/github/AyozeGS/IABD/blob/main/SNS/UT3/T1Ayoze_Gil_Sosa_Act_3_1_Metricas_de_error.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Ayoze Gil Sosa

23_24

**Actividad 3.1 - Métricas de error para los problemas de regresión**

**A partir de los ejemplos indicados realizar los ejercicios propuestos**

## **Ejercicio 1**

$$ MAE = \frac{1}{N} \sum_{i=1}^{N} |y_i - \hat{y}_i| $$

In [1]:
#Error absoluto medio - mae
from sklearn.metrics import mean_absolute_error
# Ejemplo 1
y_true = [3, -0.5, 2, 7]
y_pred = [2.5, 0, 2, 8]
mean_absolute_error(y_true, y_pred)


0.5

In [2]:
# Ejer 1: Crear dos vectores/arrays de valores ciertos y valores previstos, y calcular el mae.
# ¿Qué valores deberían de tener el array pred para que el error fuese 0? ¿y para que error fuese de 0.25?

Para que el error sea 0, los valores de ambos vectores deberían ser los mismos.

In [3]:
y_true = [10, 20, 30, 40]
y_pred = [10, 20, 30, 40]
print("MAE:", mean_absolute_error(y_true, y_pred))

MAE: 0.0


Para que el error sea 0.25, la diferencia media entre cada par de valores de ambos vectores definidos por el mismo índice debe ser 0.25.

El vector de ejemplo tiene longitud 4, por lo que si multiplicamos 0.25 por dicha longitud tenemos que la suma de todas las diferencias absolutas (es decir, eliminando el signo) debe ser igual a 1.

In [4]:
y_true = [10, 20, 30, 40]
y_pred = [10, 20, 30, 41]
print("MAE:", mean_absolute_error(y_true, y_pred))

MAE: 0.25


## **Ejercicio 2**

In [5]:
# Ejemplo 2
y_true = [[0.5, 1], [-1, 1], [7, -6]]
y_pred = [[0, 2], [-1, 2], [8, -5]]
mean_absolute_error(y_true, y_pred)

0.75

In [6]:
# Ejer 2: Crear dos matrices/arrays de valores ciertos y valores previstos, y calcular el mae.
# ¿Qué valores deberían de tener el array pred para que el error fuese 0? ¿y para que error fuese de 0.33?

Para que el error sea 0, los valores de ambas matrices deberían ser los mismos.

In [7]:
y_true = [[10, 20], [30, 40], [50, 60]]
y_pred = [[10, 20], [30, 40], [50, 60]]
print("MAE:", mean_absolute_error(y_true, y_pred))

MAE: 0.0


Para que el error sea 0.33, la diferencia media entre cada par de valores de ambas matrices definidas por los mismos índices de fila y columna debe ser 0.33.
La matriz de ejemplo tiene dimensión 2x3 y por tanto 6 elementos.

Si multiplicamos 0.33 por dicha longitud tenemos que la suma de todas las diferencias absolutas (es decir, eliminando el signo) debe ser igual a 2.


In [8]:
y_true = [[10, 20], [30, 40], [50, 60]]
y_pred = [[10, 20], [30, 40], [50, 62]]
print("MAE:", mean_absolute_error(y_true, y_pred))

MAE: 0.3333333333333333


In [9]:
y_true = [[10, 20], [30, 40], [50, 60]]
y_pred = [[10.33, 20.33], [30.33, 39.66], [50.33, 59.66]]
print("MAE:", mean_absolute_error(y_true, y_pred))

MAE: 0.33333333333333365


In [10]:
y_true = [[10, 20], [30, 40], [50, 60]]
y_pred = [[11, 20], [30, 40.33], [50.33, 59.66]]
print("MAE:", mean_absolute_error(y_true, y_pred))

MAE: 0.3333333333333333


## **Ejercicio 3**

$$ MSE = \frac{1}{N} \sum_{i=1}^{N} (y_i - \hat{y}_i)^2  $$

In [11]:
# Error cuadrático medio - mse
from sklearn.metrics import mean_squared_error
y_true = [3, -0.5, 2, 7]
y_pred = [2.5, 0.0, 2, 8]
mean_squared_error(y_true, y_pred)

0.375

In [12]:
# Ejer 3: Crear dos vectores/arrays de valores ciertos y valores previstos, y calcular el mse.
# ¿Qué valores deberían de tener el array pred para que el error fuese 0? ¿y para que error fuese de 0.25?
# Nota: Es muy interesante comparar los resultados al aplicar mae y mse sobre el mismo conjunto de datos. Podrás observar
# que mse "penaliza más" (es decir, el número que se obtiene es mayor) con las diferencias entre los valores originales y los predichos que mae


Para que el error sea 0, los valores de ambos vectores deberían ser los mismos.

In [13]:
y_true = [10, 20, 30, 40]
y_pred = [10, 20, 30, 40]
print("MSE:", mean_squared_error(y_true, y_pred))

MSE: 0.0


Para que el error sea 0.25, el cuadrado de cada diferencia entre los pares de valores de ambos vectores definidos por el mismo índice debe ser de media la raíz de dicho valor. En este caso 0.5, cuyo cuadrado corresponde a 0.25.

El vector de ejemplo tiene longitud 4, por lo que si multiplicamos 0.25 por dicha longitud tenemos que la suma de cada diferencia al cuadrado (el cuadrado elimina el signo en caso de ser negativo) debe ser igual a 1.


In [14]:
y_true = [10, 20, 30, 40]
y_pred = [10, 20, 30, 41]
print("MSE:", mean_squared_error(y_true, y_pred))

MSE: 0.25


In [15]:
y_true = [10, 20, 30, 40]
y_pred = [10.5, 20.5, 29.5, 40.5]
print("MSE:", mean_squared_error(y_true, y_pred))

MSE: 0.25


In [16]:
y_true = [10, 20, 30, 40]
y_pred = [10, 20, 29.2, 40.6]
print("MSE:", mean_squared_error(y_true, y_pred))

MSE: 0.2500000000000007


In [17]:
y_true = [10, 20, 30, 40]
y_pred = [10.7, 20.5, 30.5, 39.9]
print("MSE:", mean_squared_error(y_true, y_pred))

MSE: 0.2499999999999998


## **Ejercicio 4**

In [18]:
# Ejer 4: Crear dos matrices/arrays de valores ciertos y valores previstos, y calcular el mae.
# ¿Qué valores deberían de tener el array pred para que el error fuese 0? ¿y para que error fuese de 0.33?

Para que el error sea 0, los valores de ambas matrices deberían ser los mismos.

In [19]:
y_true = [[10, 20], [30, 40], [50, 60]]
y_pred = [[10, 20], [30, 40], [50, 60]]
print("MSE:", mean_squared_error(y_true, y_pred))

MSE: 0.0


Para que el error sea 0.33 el cuadrado de cada diferencia entre los pares de valores de ambas matrices definidas por los mismos índices de fila y columna debe ser de media la raíz de dicho valor, 0.577….

La matriz de ejemplo tiene dimensión 2x3 y por tanto 6 elementos. Si multiplicamos 0.33 por dicha longitud tenemos que la suma de cada diferencia al cuadrado (el cuadrado elimina el signo en caso de ser negativo) debe ser igual a 2.

In [20]:
y_true = [[10, 20], [30, 40], [50, 60]]
y_pred = [[10, 20], [30, 40], [51, 61]]
print("MSE:", mean_squared_error(y_true, y_pred))

MSE: 0.3333333333333333


In [21]:
y_true = [[10, 20], [30, 40], [50, 60]]
y_pred = [[10, 19.4], [30, 40.8], [49, 60]]
print("MSE:", mean_squared_error(y_true, y_pred))

MSE: 0.3333333333333328


In [22]:
y_true = [[10, 20], [30, 40], [50, 60]]
y_pred = [[10.1, 20.1], [30.5, 39.7], [50.8, 59]]
print("MSE:", mean_squared_error(y_true, y_pred))

MSE: 0.33333333333333237


## **Ejercicio 5**

$$ MAPE = \frac{100\%}{N} \sum_{i=1}^{N} \frac{|y_i - \hat{y}_i|}{|y_i|}  $$


In [23]:
# Error de porcentaje absoluto medio - mape
from sklearn.metrics import mean_absolute_percentage_error
y_true = [3, -0.5, 2, 7]
y_pred = [2.5, 0.0, 2, 8]
mean_absolute_percentage_error(y_true, y_pred)

0.3273809523809524

In [24]:
# Ejer 5: Crear dos vectores/arrays de valores ciertos y valores previstos, y calcular el mape.
# ¿Qué valores deberían de tener el array pred para que el error fuese 0?


Para que el error sea 0, los valores de ambos vectores deberían ser los mismos.

In [25]:
y_true = [10, 20, 30, 40]
y_pred = [10, 20, 30, 40]
print("MAPE:", mean_absolute_percentage_error(y_true, y_pred))

MAPE: 0.0


In [26]:
y_true = [10, 20, 30, 40]
y_pred = [10, 20, 30, 80]
print("MAPE:", mean_absolute_percentage_error(y_true, y_pred))

MAPE: 0.25


In [27]:
y_true = [10, 20, 30, 40]
y_pred = [10, 20, 45, 60]
print("MAPE:", mean_absolute_percentage_error(y_true, y_pred))

MAPE: 0.25


In [28]:
y_true = [10, 20, 30, 40]
y_pred = [12.5, 15, 37.5, 50]
print("MAPE:", mean_absolute_percentage_error(y_true, y_pred))

MAPE: 0.25


## **Ejercicio 6**

In [29]:
# Ejer 6: Crear dos matrices/arrays de valores ciertos y valores previstos, y calcular el MAPE.
# ¿Qué valores deberían de tener el array pred para que el error fuese 0?

Para que el error sea 0, los valores de ambas matrices deberían ser los mismos.

In [30]:
y_true = [[10, 20], [30, 40], [50, 60]]
y_pred = [[10, 20], [30, 40], [50, 60]]
print("MAPE:", mean_absolute_percentage_error(y_true, y_pred))

MAPE: 0.0


Para que el error sea 0.33, la diferencia media entre cada par de valores de ambas matrices definidas por los mismos índices de fila y columna debe ser un 33% del valor cierto.
La matriz de ejemplo tiene dimensión 2x3 y por tanto 6 elementos. Si multiplicamos 0.33 (33.33%) por dicha longitud tenemos que la suma de todas las diferencias absolutas (es decir, eliminando el signo) debe ser igual a 2(200%).


In [31]:
y_true = [[10, 20], [30, 40], [50, 60]]
y_pred = [[10, 20], [30, 40], [50, 180]]
print("MAPE:", mean_absolute_percentage_error(y_true, y_pred))

MAPE: 0.3333333333333333


In [32]:
y_true = [[10, 20], [30, 40], [50, 60]]
y_pred = [[20, 20], [30, 40], [50, 120]]
print("MAPE:", mean_absolute_percentage_error(y_true, y_pred))

MAPE: 0.3333333333333333


In [33]:
y_true = [[10, 20], [30, 40], [50, 60]]
y_pred = [[15, 26], [45, 60], [55, 66]]
print("MAPE:", mean_absolute_percentage_error(y_true, y_pred))

MAPE: 0.33333333333333337
